gnat2goto/driver/tree_walk.adb

```diff
@@ -2343,63 +2343,219 @@ package body Tree_Walk is
       end if;
       N_Orig := Original_Node (N);
       case Pragma_Name (N_Orig) is
-         when Name_Assert | Name_Assume | Name_Loop_Invariant =>
+         when Name_Assert |
+              Name_Assume |
+              Name_Precondition |
+              Name_Assert_And_Cut |
+            --  Assert and introduce a cut point: the prover can safely forget
+            --  evaluations of local variables and only assume the asserted
+            --  condition. This could be used in symex (making it concolic)
```

Comment:
I thought that concolic execution alternates between symex and concrete steps (selecting one evaluation) from which another symex is started. After cut point we continue with only those evaluation satisfying the assertion. It sounds similar to me.

Reply:
It's an interesting thought.  In this context, cut points are meant in the VC generation / Floyd-Hoare style verification sense.  In terms of symbolic execution this would be like throwing away the symbolic environment (havocing / non-det-ing all variables) and replacing the path condition with just the assertion.  It is an over-approximation step but it is the key thing that makes Floyd-Hoare style verification scalable.